ws/adapter.go

```diff
@@ -0,0 +1,84 @@
+package ws
+
+import (
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/gorilla/websocket"
+	"github.com/mainflux/mainflux"
+)
+
+var _ Service = (*adapterService)(nil)
+
+// Service contains publish and subscribe methods necessary for
+// message transfer.
+type Service interface {
+	mainflux.MessagePublisher
+
+	// BroadcastMessage broadcasts raw message to channel.
+	BroadcastMessage(mainflux.RawMessage)
+
+	// AddConnection adds new client ws connection for given client and channel.
+	AddConnection(IDPair, *websocket.Conn)
+}
+
+type adapterService struct {
+	pub    mainflux.MessagePublisher
+	conns  map[string]map[string]socket
+	logger log.Logger
+}
+
+// New instantiates the domain service implementation.
+func New(pub mainflux.MessagePublisher, logger log.Logger) Service {
+	return &adapterService{
+		pub:    pub,
+		conns:  make(map[string]map[string]socket),
+		logger: logger,
+	}
+}
+
+func (as *adapterService) Publish(msg mainflux.RawMessage) error {
+	as.BroadcastMessage(msg)
+	return as.pub.Publish(msg)
+}
+
+func (as *adapterService) BroadcastMessage(msg mainflux.RawMessage) {
+	chanID := msg.Channel
+	for pubID, conn := range as.conns[chanID] {
+		pid, sock := pubID, conn
```

Comment:
No need for this. If you are worried about enclosing scope of the variable, it is better to pass these as parameters to the goroutine.